File: project-01-rbac-system/code-java/src/main/java/com/example/demo/repository/PermissionRepository.java
package com.example.demo.repository;

import org.jooq.DSLContext;
import static org.jooq.impl.DSL.selectOne;
import org.springframework.stereotype.Repository;

import static com.example.jooq.tables.Permission.PERMISSION;
import static com.example.jooq.tables.RolePermission.ROLE_PERMISSION;
import static com.example.jooq.tables.UserRole.USER_ROLE;

@Repository
public class PermissionRepository {
    private final DSLContext dsl;

    public PermissionRepository(DSLContext dsl) {
        this.dsl = dsl;
    }

    public boolean hasPermission(Long userId, String resource, String action) {
        return dsl.fetchExists(
            selectOne()
            .from(USER_ROLE)
            .join(ROLE_PERMISSION).on(USER_ROLE.ROLE_ID.eq(ROLE_PERMISSION.ROLE_ID))
            .join(PERMISSION).on(ROLE_PERMISSION.PERMISSION_ID.eq(PERMISSION.ID))
            .where(USER_ROLE.USER_ID.eq(userId))
              .and(PERMISSION.RESOURCE.eq(resource))
              .and(PERMISSION.ACTION.eq(action))
        );
    }
}

File: project-01-rbac-system/code-kotlin/src/main/kotlin/com/example/demo/repository/PermissionRepository.kt
package com.example.demo.repository

import com.example.jooq.tables.Permission.PERMISSION
import com.example.jooq.tables.RolePermission.ROLE_PERMISSION
import com.example.jooq.tables.UserRole.USER_ROLE
import org.jooq.DSLContext
import org.jooq.impl.DSL.selectOne
import org.springframework.stereotype.Repository

@Repository
class PermissionRepository(private val dsl: DSLContext) {

    fun hasPermission(userId: Long, resource: String, action: String): Boolean {
        return dsl.fetchExists(
            selectOne()
            .from(USER_ROLE)
            .join(ROLE_PERMISSION).on(USER_ROLE.ROLE_ID.eq(ROLE_PERMISSION.ROLE_ID))
            .join(PERMISSION).on(ROLE_PERMISSION.PERMISSION_ID.eq(PERMISSION.ID))
            .where(USER_ROLE.USER_ID.eq(userId))
              .and(PERMISSION.RESOURCE.eq(resource))
              .and(PERMISSION.ACTION.eq(action))
        )
    }
}
